File: ml-class/ex4/predict_nn.cpp
/*
 *  Predict usando o Armadillo para C++
 *
 *   g++ -ggdb -o predict_nn predict_nn.cpp -O1 -larmadillo
 *
 *
 *
 */


#include <iostream>
#include <armadillo>

using namespace std;
using namespace arma;

int main(int argc, char** argv)
{

	/*
	 // exemplo
	mat A (2,2);

	A(0,0) = 2;
	A(0,1) = 3;
	A(1,0) = 4;
	A(1,1) = 5;

	//mat O = ones<mat> (2,1);
	//A.insert_cols(0,O);

	//cout << A << endl;

	//cout << exp(-A) << endl;
	// fim do exemplo
	*/

	mat Exemplo;
	mat Theta1, Theta2;

	Exemplo.load("/home/tercio/projects/tercio/Machine-Learning/ml-class/ex4/exemplo.mat",raw_ascii); 
	Theta1.load("/home/tercio/projects/tercio/Machine-Learning/ml-class/ex4/Theta1.mat",raw_ascii); 
	Theta2.load("/home/tercio/projects/tercio/Machine-Learning/ml-class/ex4/Theta2.mat",raw_ascii); 

	//cout << Exemplo;

	// sigmoid function
	// g = 1.0 ./ (1.0 + exp(-z));	


	int m=1;
	int num_labels = 10;


	Row<double> Ones = ones<vec> (1);

	Exemplo.insert_cols(0,Ones);
	Row<double> h1 = 1.0 / (1.0 + exp(-(Exemplo * trans(Theta1) )));
	//cout << "h1-> " << h1.n_rows << " x " << h1.n_cols<<endl;

	//cout << h1 << endl;


	h1.insert_cols(0,Ones);
	mat h2 = 1.0 / (1.0 + exp(-(h1 * trans(Theta2) )));
	//cout << "h2-> " << h2.n_rows << " x " << h2.n_cols<<endl;

	cout << h2<<endl;
	uword index;
   	double val = h2.max(index);

	index ++;
	cout << "ans: " << index % 10 << "  (index: " << index<< ") (val: " << val << ")" << endl;

	return 0;
}

File: ml-class/ex4/predict_nn.py
#!/usr/bin/python

import numpy as np
import scipy.io
import cv2

def sigmoid (z):
	'''SIGMOID Compute sigmoid functoon
	      J = SIGMOID(z) computes the sigmoid of z.'''



	#     g = 1.0 ./ (1.0 + exp(-z));	
	return 1.0 / (1.0 + np.exp( -z ))


if __name__ == "__main__":

	np.seterr (over='ignore')

	# le uma imagem. Nesse caso deve ser uma imagem com um digito igual ao mnist de 20x20
	im = cv2.imread('teste4.png', cv2.CV_LOAD_IMAGE_GRAYSCALE)

	im = im.flatten(1) # same as im(:) in octave

	# carrega os Thetas gerados pelo treinamento da Rede Neural no Octave 
	# o valor aqui e um dict
	Theta1 = scipy.io.loadmat ('./theta1.mat')
	Theta2 = scipy.io.loadmat ('./theta2.mat')

	# converte os Thetas para ndarray
	Theta1 = Theta1['Theta1']
	Theta2 = Theta2['Theta2']

	# qtde de inputs (nesse caso so uma imagem)
	m=1;
	# camada de saida: dez unidades (digitos 0 - 9)
	num_labels = 10;

	# feed forward camada entrada -> camada escondida
	Exemplo = np.concatenate ([np.ones(1),im],axis=1)
	# sigmoid (z) : z e o valor de a* Theta' 
	h1 =  sigmoid (np.dot(Exemplo , Theta1.transpose()) )

	# feed forward camada escondida -> camada de saida
	h1 = np.concatenate ([ np.ones(1), h1 ]);
	h2 =  sigmoid ( np.dot(h1 , Theta2.transpose()) )

   	val = h2.max()
	# na lista de outputs, qual o que tem maior peso?  esse e o digito encontrado
	index = h2.argmax()

	# mostra os dados. Note que o index e incrementado, pois comecamos no digito 1 e o zero foi transformado para 10
	# para compatibilidade e maior facilidade com o Octave, que tem indice de array inicial igual a 1
	index = index + 1
	print ("ans: %d  (index: %d) (val: %f)\n" % (index%10,index,val) )


